Protecting Your Assets with Replicated (#327)
replicatedhq/replicated-field-labs: instruqt/protecting-your-assets/03-proxying-private-images/assignment.md

type: challenge
title: Proxying Private Images
teaser: Protect your private container images with the Replicated proxy registry
notes:
- type: text
  contents: |
    Share your private images without exposting your private registry to your customers
tabs:
- title: Vendor Portal
  type: website
  url: https://vendor.replicated.com
  new_window: true
- title: Slackernews Chart
  type: code
  hostname: shell
  path: /home/replicant
- title: Shell
  type: terminal
  hostname: shell
  workdir: /home/replicant
difficulty: basic
timelimit: 600
---

One of the core features of the Replicated Platform is it's proxy registry. The
proxy registry controls access to your images using the Replicated license
This relieves you of the burden of managing authentication and authorization
for the private images your application depends on. You provide Replicated with
access and we manage the rest.

Configuring a Private Registry
==============================

The first step in using the Replicated proxy registry is to provide access to
your private registry for the Replicated Vendor Portal. This has already been
done in the lab environment, so we're just going to review how it was set.

> Log into the Vendor Portal with the following credentials:

Reviewer: ```suggestion
If you logged out between challenges, log back into the Vendor Portal with the following credentials:
```

Author: Reworded a little bit differently.

> 

Reviewer: I was this many days old when my instruct expired. Maybe I'm just slow, but this seems like an aggressively low timeout is that something you can increase?

Reviewer: Oh crap I have to re-create my customer when that happens too!

Author: I've definitely increased the timeout. Chatting with @xavpaice about whether it's enough.

Reviewer 2: Took me an hour to go over it in some depth, and if I was totally new to everything it might be double that.

From the "Channels" page you landed on, select "Images" in the left-hand menu.
You'll land on a page that shows that the proxy is configured to pull private
images from the GitHub Container Registry.

![Managing Images on the Vendor Portal](../assets/managing-images.png)

You use the "+ Add external registry" button to add a private registry. We won't
be doing this in the lab environment since it's shared team.

Adding the Proxy to Your Chart
==============================

The next step to securing your images with the Replicated Platform is to use
the proxy to pull images. The Slackernews chart refers to its two images in
it's default values. We'll change those to refer to the Replicated proxy
registry.

Images tags in the proxy registry follow the following pattern:

```
$PROXY_REGISTRY_HOST/proxy/$APP_SLUG/$ORIGINAL_IMAGE_TAG
```

so you'll proxy the image `ghcr.io/slackernews/slackernews-web:1.0.17` as

```
proxy.replicated.com/proxy/[[ Instruqt-Var key="REPLICATED_APP" hostname="shell" ]]/ghcr.io/slackernews/slackernews-web:1.0.17
```

for your instance of Slackernews. Let's do that now.

### Updating Image References

> Select the "Slackernews Chart" tab to edit the chart source files. Select the

Reviewer: There isn't a tab for this? I think I found it under Releases -> Active Releases -> View Yaml -> values.yaml in the left side bar?

Reviewer: ```suggestion
Select the "Slackernews Chart" tab in Instruqt to edit the chart source files. Select the
```

OH this is talking about the tab in instruqt not in vendor portal! I compeletely missed that.

file `values.yaml` and search for the image values.

![Image Value Settings in Helm Chart](../assets/image-values.png)

Update the image values to use the proxied image for the main Slackernews
image

```
slackernews:
  repository: proxy.replicated.com/proxy/[[ Instruqt-Var key="REPLICATED_APP" hostname="shell" ]]/ghcr.io/slackernews/slackernews-web:1.0.17
```

Since we've made a change that will impact our users, let's also bump the
version number by editing `Chart.yaml` and setting a new chart version.

```

> version: 0.5.0

Reviewer: Shouldn't this warrant a patch not a minor change? ;)

Author: The pattern I've been using is "new lab is new minor" because we're using a new set of platform capabilities.

```

The editor should have automatically saved your changes to both files.

Allowing Image Access
---------------------

The last change you need to make is to assure an appropriate image pull secret
is set on the pods that require access to proxied images. We need two things to
do this:

1. A template in our Helm chart to create the secret
2. Reference to the secret for any workloads that need it

Another good practice is to specify image pull secrets as Helm values, so we'll
plan on doing that as well.

> The image pull secret provides registry credentials to pull image. An obvious

Reviewer: ```suggestion
The image pull secret provides registry credentials to pull images. An obvious
```

question to ask is how we know which credentials to use for the proxy registry.

> Replicated takes care of this for is as part of the same process it uses to

Reviewer: ```suggestion
Replicated takes care of this for us as part of the same process it uses to
```

inject the license file into your Helm chart. This process provides a global
value `replicated.dockerconfigjson` that contains the credentials.

We're going to go through the changes above backwards, since Slackernews has
already taken care of the second one. Open the `slackernews-deployment.yaml`
file under templates and look for the following lines that use the image pull
secret.

```
      {{ if hasKey ((.Values.global).replicated) "dockerconfigjson" }}
      imagePullSecrets:
        - name: replicated-pull-secret
```

> This lines are only included if the global values is set, in other

Reviewer: ```suggestion
These lines are only included if the value `global.repliacted` is set, in other
```

Author: I'd like to keep this descriptive but it wasn't worded correctly. I'll update the way I'm thinking.

your shell environment. The lab setup has created a token for you.

```
export REPLICATED_API_TOKEN="[[ Instruqt-Var key="REPLICATED_API_TOKEN" hostname="shell" ]]"
```

The `replicated` command also needs to know which application to work with. You
can set it with a command flag or with an environment variable. Let's set the
variable to save some typing.

```
export REPLICATED_APP="[[ Instruqt-Var key="REPLICATED_APP" hostname="shell" ]]"
```

To create a new release, run the following command.

```
replicated release create --promote Unstable --chart ./release/slackernews-0.5.0.tgz --version 0.5.0  \
  --release-notes "Uses the Replicated proxy registry to pull images"
```

This will not only create the release but push it to the `Unstable` release
channel on the platform. It will also output a sequence number, which should be
`5` (if it's not, use the number in the output for the upcoming commands).
We're going to release our changes on three different release channels in this
lab:

* `Unstable`: use for internal releases that aren't ready for customers
* `Beta`: for releases that are ready for beta testing
* `Stable`: the general availability channel for all releases

> 

Reviewer: I also got a comment from the CLI that I was creating a helm only release. Might be nice to let people know this same release can be extended to KOTS/Embedded Cluster this isn't work that is only for helm cli.

Author: Good point.

---
slug: proxying-private-images
id: ui20iijaawgo
type: challenge
title: Proxying Private Images
teaser: Protect your private container images with the Replicated proxy registry
notes:
- type: text
  contents: |
    Share your private images without exposting your private registry to your customers
tabs:
- title: Vendor Portal
  type: website
  url: https://vendor.replicated.com
  new_window: true
- title: Slackernews Chart
  type: code
  hostname: shell
  path: /home/replicant
- title: Shell
  type: terminal
  hostname: shell
  workdir: /home/replicant
difficulty: basic
timelimit: 600
---

One of the core features of the Replicated Platform is it's proxy registry. The

> proxy registry controls access to your images using the Replicated license

Reviewer 2: ```suggestion
proxy registry controls access to your images using the Replicated license.
```
